README.md

```diff
@@ -0,0 +1,400 @@
+        Cell::new(self.into_inner().coerce_unsized())
+    }
+}
+```
+
+A delegating impl is not limited to struct types, as such another example implementation for `Option<T>` would be:
+```rs
+impl<T, U> CoerceUnsized<Option<U>> for Option<T>
+where
+    T: CoerceUnsized<U>,
+{
+    fn coerce_unsized(self) -> Option<U> {
+        match self {
+            Option::Some(t) => Option::Some(t.coerce_unsized()),
+            Option::None => Option::None,
+        }
+    }
+}
+```
+With such an impl, `Option<&[T; N]>` could coerce to `Option<&[T]>`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+
+## `Unsize`
+
+The new `Unsize` trait definition looks like the following:
+
+```rs
+/// # Safety
+///
+/// The implementation of [`Unsize::target_metadata`] must return metadata that
+/// - is valid for interpreting the `Self` type to `Target`, and
+/// - where using `core::mem::size_of` on the unsized object will report the same size as on the source object.
+pub unsafe trait Unsize<Target>
+where
+    Target: ?Sized,
+{
+    fn target_metadata(metadata: <Self as Pointee>::Metadata) -> <Target as Pointee>::Metadata;
+}
+```
+
+This trait allows specifying how to derive the metadata required for unsizing from the metadata of the source type `Self` or the compile time type information.
+
+## `CoerceUnsized`
+
+The new `CoerceUnsized` trait definition looks like the following:
+
+```rust
+pub trait CoerceUnsized<Target> {
+    fn coerce_unsized(self) -> Target;
+}
+```
+
+Implementations of this trait now specify how the coercion is done.
+This also drops the `?Sized` bound on `Target`, as returning unsized values is not possible currently and it being possible to relax the bound later on.
+
+In order to prevent misuse of the trait as means of implicit conversions, implementations for this trait require specific conditions to hold which the compiler will enforce. This can be relaxed without breakage in the future.
+For an implementation to be valid, one of the following must hold:
+1. `Self` and `Target`
+    - must be references or raw pointers to different generic parameters
+    - type parameter `T` of `Self` has `T: Unsize<U>` bound where `U` is the type parameter of `Target`
+2. `Self` and `Target`
+    - must have the same type constructor, varying in a single type parameter
+    - type parameter `T` of `Self` must have a `T: CoerceUnsized<U>` bound where `U` is the type parameter of `Target`
+    - Example:
+        ```rust
+        impl<T: CoerceUnsized<U>, U> CoerceUnsized<Cell<U>>
+            for Cell<T>
+        ```
+3. `Self` and `Target`
+    - must have the same type constructor, varying in a single type parameter
+    - type parameter `T` of `Self` must have a `T: Unsize<U>` bound where `U` is the differing type parameter of `Target`
+    - Example:
+        ```rust
+        impl<T: ?Sized + Unsize<U>, U: ?Sized, A: Allocator> CoerceUnsized<Box<U, A>>
+            for Box<T, A>
+        ```
+
```

Comment:
I feel like case two and three should be merged since they are largely similar

Reply:
That was the case before, aman proposed to split them as that makes it more distinguishable.